tests/Unit/Helpers/ParallelAlgorithms/LinearSolver/Multigrid/Helpers.hpp

```diff
@@ -0,0 +1,252 @@
+struct ComputeOperatorAction {
+  using const_global_cache_tags = tmpl::list<LinearOperator>;
+  using local_operator_applied_to_operand_tag =
+      db::add_tag_prefix<::LinearSolver::Tags::OperatorAppliedTo, OperandTag>;
+
+  template <typename DbTagsList, typename... InboxTags, typename Metavariables,
+            typename ActionList, typename ParallelComponent>
+  static std::tuple<db::DataBox<DbTagsList>&&, Parallel::AlgorithmExecution>
+  apply(db::DataBox<DbTagsList>& box,
+        const tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+        const Parallel::GlobalCache<Metavariables>& /*cache*/,
+        const ElementId<1>& element_id, const ActionList /*meta*/,
+        const ParallelComponent* const /*meta*/) noexcept {
+    const size_t mg_lev = element_id.grid_index();
+    const size_t element_index = helpers_distributed::get_index(element_id);
+    const auto& operator_matrices = get<LinearOperator>(box)[mg_lev];
+    const size_t number_of_elements = operator_matrices.size();
+    const auto& linear_operator = gsl::at(operator_matrices, element_index);
+    const size_t number_of_grid_points = linear_operator.columns();
+    const auto& operand = get<OperandTag>(box);
+
+    typename OperandTag::type operator_applied_to_operand{
+        number_of_grid_points * number_of_elements};
+    dgemv_('N', linear_operator.rows(), linear_operator.columns(), 1,
+           linear_operator.data(), linear_operator.spacing(), operand.data(), 1,
+           0, operator_applied_to_operand.data(), 1);
+
+    auto& section = *db::get_mutable_reference<Parallel::Tags::Section<
+        ParallelComponent, ::LinearSolver::multigrid::Tags::MultigridLevel>>(
+        make_not_null(&box));
+    Parallel::contribute_to_reduction<CollectOperatorAction<OperandTag>>(
+        Parallel::ReductionData<
+            Parallel::ReductionDatum<typename OperandTag::type, funcl::Plus<>>,
+            Parallel::ReductionDatum<size_t, funcl::AssertEqual<>>>{
+            operator_applied_to_operand, mg_lev},
+        section.proxy()[element_id], section.proxy(), make_not_null(&section));
+
+    // Pause algorithm for now. The reduction will be broadcast to the next
+    // action which is responsible for restarting the algorithm.
+    return {std::move(box), Parallel::AlgorithmExecution::Pause};
+  }
+};
+
+template <typename OperandTag>
+struct CollectOperatorAction {
+  using local_operator_applied_to_operand_tag =
+      db::add_tag_prefix<::LinearSolver::Tags::OperatorAppliedTo, OperandTag>;
+
+  template <typename ParallelComponent, typename DbTagsList,
+            typename Metavariables, typename ScalarFieldOperandTag,
+            Requires<tmpl::list_contains_v<
+                DbTagsList, local_operator_applied_to_operand_tag>> = nullptr>
+  static void apply(
+      db::DataBox<DbTagsList>& box, Parallel::GlobalCache<Metavariables>& cache,
+      const ElementId<1>& element_id,
+      const Variables<tmpl::list<ScalarFieldOperandTag>>& Ap_global_data,
```

Comment:
What does the `Ap` prefix mean in this name?

Reply:
It means the linear operator `A` applied to the operand `p`, i.e. `A \cdot p`. Changed to `operator_applied_to_operand`.